compiler/frontend/beam/src/main/java/org/apache/nemo/compiler/frontend/beam/source/BeamBoundedWindowSourceVertex.java

```diff
@@ -0,0 +1,148 @@
+
+/**
+ * SourceVertex implementation for BoundedSource.
+ * @param <O> output type.
+ */
+public final class BeamBoundedWindowSourceVertex<O> extends SourceVertex<WindowedValue<O>> {
+  private static final Logger LOG = LoggerFactory.getLogger(BeamBoundedWindowSourceVertex.class.getName());
+  private BoundedSource<O> source;
+  private final String sourceDescription;
+
+  /**
+   * Constructor of BeamBoundedWindowSourceVertex.
+   *
+   * @param source BoundedSource to read from.
+   */
+  public BeamBoundedWindowSourceVertex(final BoundedSource<O> source) {
+    super();
+    this.source = source;
+    this.sourceDescription = source.toString();
+  }
+
+  /**
+   * Constructor of BeamBoundedWindowSourceVertex.
+   *
+   * @param that the source object for copying
+   */
+  public BeamBoundedWindowSourceVertex(final BeamBoundedWindowSourceVertex that) {
+    super(that);
+    this.source = that.source;
+    this.sourceDescription = that.source.toString();
+  }
+
+  @Override
+  public BeamBoundedWindowSourceVertex getClone() {
+    return new BeamBoundedWindowSourceVertex(this);
+  }
+
+  @Override
+  public List<Readable<WindowedValue<O>>> getReadables(final int desiredNumOfSplits) throws Exception {
+    final List<Readable<WindowedValue<O>>> readables = new ArrayList<>();
+    LOG.info("estimate: {}", source.getEstimatedSizeBytes(null));
+    LOG.info("desired: {}", desiredNumOfSplits);
+    source.split(source.getEstimatedSizeBytes(null) / desiredNumOfSplits, null)
+        .forEach(boundedSource -> readables.add(new BoundedSourceReadable<>(boundedSource)));
+    return readables;
+  }
+
+  @Override
+  public void clearInternalStates() {
+    source = null;
+  }
+
+  @Override
+  public ObjectNode getPropertiesAsJsonNode() {
+    final ObjectNode node = getIRVertexPropertiesAsJsonNode();
+    node.put("source", sourceDescription);
+    return node;
+  }
+
+  /**
+   * BoundedSourceReadable class.
+   * @param <T> type.
+   */
+  private static final class BoundedSourceReadable<T> implements Readable<WindowedValue<T>> {
+    private final BoundedSource<T> boundedSource;
+
+    /**
+     * Constructor of the BoundedSourceReadable.
+     * @param boundedSource the BoundedSource.
+     */
+    BoundedSourceReadable(final BoundedSource<T> boundedSource) {
+      this.boundedSource = boundedSource;
+    }
+
+    @Override
+    public Iterable<WindowedValue<T>> read() throws IOException {
+      boolean started = false;
+      boolean windowed = false;
+
+      final ArrayList<WindowedValue<T>> elements = new ArrayList<>();
+      try (BoundedSource.BoundedReader<T> reader = boundedSource.createReader(null)) {
+        for (boolean available = reader.start(); available; available = reader.advance()) {
+          final T elem = reader.getCurrent();
+
+          if (!started) {
+            started = true;
+            if (elem instanceof WindowedValue) {
+              windowed = true;
+            }
+          }
+
+          if (!windowed) {
```

Comment:
I'd do away with `started` and `windowed`, and do here:
if (elem instanceof WindowedValue)

Reply:
I thought this may lead to a better performance because we don't have to check all elements (it is enough to check the first element)